Fail on `EC-SIZE` exceptions as per the standard

emilienlemaire opened the pull request.

emilienlemaire committed:
Handle size exceptions as per the standard with `-fstandard-exception`

emilienlemaire committed:
Add changelog messages

GitMensch commented:
Why do we need a new flag instead of -fec?
Note: if not implemented yet we should add the cobol2027 ec-all-fatal to enable only those checks (that needs to iterate over the existing table instead of the current "enable the single/group entry".

Note: no changelog needed for new tests

emilienlemaire commented:
The new flag is introduced because the current behavior is to continue normally when fatal exceptions are raised and are not checked, which is not the behavior described by the standard.

To avoid breaking programs running on GC3 after the PR is merged I added the flag, so now user can choose how these exceptions should behave, i.e do nothing as the default behavior or fail with the flag set.

GitMensch commented:
I see but that's exactly what -fec is for: giving fine grained control about which exceptions are checked. We have a known defect that -fec=size does not work in every case and that will be fixed with this PR being finished.
If people compile with -fec=all or --debug and don't want the the working check, they can just add -fno-ec=size

(But we only got multiple people reporting that this does not abort, even with those runtime checks enabled).
Emilien Lemaire ***@***.***> schrieb am Dienstag, 7. Oktober 2025 um 20:35:

> emilienlemaire left a comment (OCamlPro/gnucobol#247)
> 

> The new flag is introduced because the current behavior is to continue normally when fatal exceptions are raised and are not checked, which is not the behavior described by the standard.
> 

> To avoid breaking programs running on GC3 after the PR is merged I added the flag, so now user can choose how these exceptions should behave, i.e do nothing as the default behavior or fail with the flag set.
> 

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.

emilienlemaire committed:
Remove `-fstandard-exception`, and use `-fno-ec=EC-SIZE`

emilienlemaire commented:
I edited the tests that shouldn't work with `-fec=EC-SIZE` to add `-fno-ec=EC-SIZE` so they pass

GitMensch commented:
While I'll need to check the details (@ddeclerck can possibly help with the all failing CI tests):
* tests/Changelog can be reverted
* `EC-SIZE-EXPONENTIATION` should be included
* for the testcases I'd suggest (but not enforce) using the shortened form ` fno-ec=SIZE` (possibly lower-case)
* run_misc should include a test for all checked `EC-SIZE` - that can go all into a single testcase with multiple programs + `AT_CHECK`s
* NEWS should have a user-view entry for this change

emilienlemaire committed:
Update tests

emilienlemaire commented:
I added the tests for all the exceptions that are set in libcob

emilienlemaire committed:
Add exception failure to config

emilienlemaire commented:
I also added a configuration option to enable the failure of the exceptions. I figured many could want to run with `--debug`, yet the COBOL dialect they used should not fail on such exceptions, and this avoid adding `-fno-ec` each time it is the case

GitMensch commented on cobc/config.def:
we don't need the (currently not documented) cb_fatal_exceptions_fail, so can revert this file and all of config - we already have -fec; in a follow-up PR we can consider adding "default" exceptions per language _if_ we find a need, but I think apart from PACKAED-DECIMAL abends on IBM there's like no default exception.

emilienlemaire replied:
So should we add `-fno-ec=size` to all the NIST tests? (if yes where should I do that, I get lost in all the config files from automake and stuff)?

GitMensch replied:
That's interesting, you say that those fail?  And do they with all size exceptions or only with overflow?
To answer your question: that's `COBC_FLAGS` in tests/cobol85/Makefile.am directly after what _should_ be `--debug` [there's currently only one hyphen so a good thing to fix at the same time].

emilienlemaire replied:
Some of them do, around 10 or 20 if i recall correctly, I'll test again without the new config to get the exact number

GitMensch commented on cobc/codegen.c:
```diff
@@ -9150,6 +9205,12 @@ output_stmt (cb_tree x)
+		if ((!p->ex_handler || (p->ex_handler && p->handler_type != SIZE_ERROR_HANDLER))
+			&& cb_fatal_exceptions_fail) {
+			output_size_exception_handler(p);
+		}
```
Is there a very good reason to add that after the tree is finished?
Wouldn't `cb_emit_arithmetic()` be the right place?

emilienlemaire replied:
We don't have info about the exception handler in `cb_emit_arithmetics()`, so this would mean adding this information in the function, which I am not sure we want to do, and the checks need to be done after the statement, not during intermediate calculations, where the exceptions are to be set, but not checked yet.

GitMensch commented on cobc/codegen.c:
```diff
@@ -9033,6 +9033,61 @@ output_assign (const struct cb_assign *ap)
+					output_line ("cob_fatal_error(COB_FERROR_FATAL_EC);");
+					output_block_close ();
+					output_block_close ();
+				}
```
Those exceptions are all fatal - is there a reason to not use the current way of checking by adding a call to a new function `cob_call_size_error (enabled_exceptions)` as error handler if no `[NOT] ON SIZE ERROR` is coded?

emilienlemaire replied:
I wanted to this first, but we don't have a lot of info about enabled exceptions at runtime, and I figured it would be more efficient to generate a switch at compile time than having a runtime function check for all the enabled exceptions every time it is called

GitMensch replied:
The exception handler will only be generated _if_ the exception was active - and you can do that with a switch-case statement similar to what you did as well.
Side note: the codegen is a _bit_ special here for the other runtime checks as codegen.c has the direct function call but codeoptim.def optimizes most of them to an inline call which is verySoonTM (when I've split that from other local changes) changed to a define instead. The point for this was: you can fast-read the generated C text which has a "direct function call" for the check (also: diffs to older versions are identical), but the actual code via C preprocessor does a "local" check before calling into libcob for no reason.

For this new check of yours the `switch (cob_glob_ptr->cob_exception_code)` for the specific cases as an *implicit* `ON SIZE ERROR` if no handler is coded at all is the good way for the new check. I haven't checked but you may add that as `on_exception` tree from a `cb_build_direct` or similar (doesn't look that pretty in the generated C but I don't mind).

emilienlemaire replied:
Sorry but I don't really understand what you mean or ask for here (I am not that knowledgeable on codegen yet, but getting there :sweat_smile: )...

Here was my trail of thoughts for implementing the checks this way:

What I meant to do was to keep the code as close as possible as if we had a `ON SIZE ERROR` (the `NOT ON` does not really matter in our case, as even if it is present we need to check for exceptions that were set), or if you will as if the default was an implicit `ON SIZE ERROR` that terminates the program in case of an exception. So I based my code from what the `ON SIZE ERROR` codegen did (and also where it was).

Then I wondered how to make it so there was no code generated if not needed, and how to make it as efficient as possible, because I though this part might be executed quite a lot during a run of any program. So I thought having a switch that only check for the activated exception codes. But I also saw that `COB_EC_SIZE_IMP` (although never set in libcob) is not fatal, so I just added a check for fatal errors.

(I made this recap so you can see why I did it this way)

Having written my recap and re-reading your message, I think I understand what you meant better. Instead of the switch, you'd like me to make a function `cob_call_size_error (enabled_exceptions)` (which would still mean having enabled exceptions registered somewhere to be used at runtime), which can be optimized away using `codeoptim.def` when needed, this codeoptim could itself have a generated switch which would be filled accordingly with enabled exceptions, and then there wouldn't be any need for runtime checks of enabled exceptions?

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -14039,7 +14039,7 @@ AT_SETUP([Constant Expressions])
 
 AT_DATA([prog.cob], [
-       IDENTIFICATION DIVISION.
+       IDENTIFICATION DIVISION
```
that looks like an error

emilienlemaire replied:
Yes I missed it when reading my diffs :sweat_smile:

GitMensch commented on tests/testsuite.src/run_misc.at:
Can you please make the suppressed exception check as specific as possible? I guess that in most cases this will be `-fno-ec=size-overflow`.

... second thought: Most likely the better move in most cases is to add an `ON SIZE ERROR CONTINUE` to those tests.

emilienlemaire replied:
I thought about adding the `ON SIZE ERROR`, but in my mind it would change the semantics of the tests which is not something I wanted to do

GitMensch replied:
Depends on the test -  the ones that do want to check for the exception by the exception functions _should_ have been coded with `ON SIZE ERROR CONTINUE` in any case.
But for the review both suppressing the (specific as possible) check or a the conditional `CONTINUE` or a combination are all fine.

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -15682,3 +15682,176 @@ AT_CHECK([$COBCRUN_DIRECT ./prog6], [0],
+], [])
+
+
+AT_CLEANUP
```
please add a trailing newline

GitMensch commented on libcob/ChangeLog:
```diff
@@ -1,3 +1,13 @@
+2025-10-07  Emilien Lemaire <emilien.lemaire@ocamlpro.com>
```
please add a leading newline instead of the double-trailing one

GitMensch reviewed (changes requested):
I think we getting near to the solution :-)

GitMensch commented:
> I also added a configuration option to enable the failure of the exceptions. I figured many could want to run with `--debug`, yet the COBOL dialect they used should not fail on such exceptions, and this avoid adding `-fno-ec` each time it is the case

`--debug` is something similar (actually more) that you have with other compiler switches on other compilers where you need to enable checks as well. It is fine to keep as-is and if someone ports he can have a one-time conversion of his old options to the new ones (or let a consultant do that or even better, let someone write a wrapper that takes the old command line options (and possibly environment variables) and calls cobc with the appropriate options).

emilienlemaire committed:
Update tests and config

emilienlemaire committed:
Generate exception handler in tree

emilienlemaire commented:
Hello @GitMensch,
I am trying to implement what I think you asked for in [this comment](https://github.com/OCamlPro/gnucobol/pull/247#discussion_r2417255380), but as I mentioned it is not that easy, as we don't have knowledge of what exceptions have been disabled at runtime.

I made [this commit](https://github.com/OCamlPro/gnucobol/pull/247/commits/25ef5e1336c0888b3aed311964904f753bde3204) (which will be edited), and the runtime knowledge is the last part I have to add, so I was wondering how I should add the information on the exceptions that enabled / disabled ?

GitMensch commented:
👋 Emilien and thanks again for working on this.

Concerning your quuestion - the information about the checked exceptions is codegen'd into the module - it can even be different "per line" using the TURN directives.
The runtime just always sets the specific exception code and the generated code checks per "what is active":

* if there are tests<, then before the actual statement the same value that is tested for later on (normally checked by its group) is reset
* then the code in the runtime is executed which would set those values
* then the exception values actually active on this line will have a generated test for

that approach _should_ be visible if you now generate with `ON SIZE ERROR` - it "just" needs to be added for the error case without that clause

Your last commit seems to do that quite cleanly, no?

How does the generated code with for example a check for divzero look like?

Did you do anything to the conditions (`IF MY-VAR / MY-OTHER-VAR` may create a divzero) as well or is that open?

What are the parts this PR needs a direction / review on?

Note: instead of a `cob_call_size_error (void)` it seems better top have something more general like `cob_fatal_exception (int exception_code)`.

emilienlemaire commented:
Hello Simon!

Thank you for the help!

I updated my commits, and hopefully we are getting closer to get it merged :smile: 

Concerning the `IF MY-VAR / MY-OTHER-VAR` I have not yet added any code that can handle that, I'll look into it.

I think I made the changes you were looking for, yet this pose a problem: when any `EC-SIZE` exception is enabled then the check is valid for all `EC-SIZE` errors, this was something that was avoided with my initial `switch` generation. This can also be avoided by generating a check for enabled error at compile time, but I don't think it would be a good idea to generate this inside `parser.y`, so this would mean moving the code generation to codegen.c and do something closer to what I did before (excluding the configs and flags I added, which we can forget about).

GitMensch commented:
Why do the changes here fail one of the NC tests? did the behavior changed between standards (and we miss a possible distinction between COBOL85 and 2002+) or is there an error in the NIST test or does it depend on undefined behaviour ... or there is an error in the PR itself?

Note: In _all_ cases: please try to "extract" the failure into one of the tests in our own testsuite before working on fixing it,
Thanks for your efforts in this important area (and note that this is one of the missing changes we need this year before the 3.3rc and after that we want to sent the final out asap [=if no breaking tests occur within 2- weeks]).

> Concerning the IF MY-VAR / MY-OTHER-VAR I have not yet added any code that can handle that, I'll look into it.

Thanks. In effect it should have a "post check" generated that inspects the error code - similar as the generation without the SIZE ERROR clause.

For the EC-SIZE error problem - that's no issue as long as the check either:
* generates specific check for the enabled EC-SIZEs
* generates a check that will go into the runtime for any EC-SIZE and passes the EC-SIZE values needed into the function

Just a fair warning: I'm likely not be available for any (or any longer) reviews this week (and also did not checked the current changes either, just the CI results).

emilienlemaire committed:
Update NIST test compilation options

emilienlemaire commented:
There was no `EXCEPTION` in the COBOL85 standard, and I forgot to update the compilation options in my commit, hence the NIST failures (which are remediated by the last commit).

emilienlemaire committed:
Generate per exception tests

emilienlemaire committed:
Revert wrongly committed files

GitMensch commented on ABOUT-NLS:
change should be reverted, the current version is the replacement link (the file is checked in our current minimal autoconf for its availability so we use that one)

GitMensch replied:
please commit the resolving change

GitMensch commented on NEWS:
```diff
@@ -6,6 +6,19 @@ NEWS - user visible changes				-*- outline -*-
+** at runtime, fatal size exception codes now fail when activated (with
+   `-fec=ALL` or `-fec=SIZE or its subordinates). Theses exceptions can be
+   caught with `ON SIZE ERROR` phrases to avoid the program failing when
+   they are set. The affected exceptions codre are:
```
typo, should be: "exception codes"

GitMensch commented on NEWS:
```diff
@@ -6,6 +6,19 @@ NEWS - user visible changes				-*- outline -*-
 * New GnuCOBOL features
 
+** at runtime, fatal size exception codes now fail when activated (with
+   `-fec=ALL` or `-fec=SIZE or its subordinates). Theses exceptions can be
```
not sure if we use backticks in that file, if yes, then the closing one for ec=SIZE is missing
As this paragraph is a sentence it should start uppercase

GitMensch commented on cobc/ChangeLog:
```diff
@@ -1,3 +1,10 @@
+2025-11-10  Emilien Lemaire <emilien.lemaire@ocamlpro.com>
```
one empty line before and after the Changelog entry, please (currently it is 0/2) and indent continuation lines by two spaces

GitMensch commented on cobc/codegen.c:
```diff
@@ -8718,6 +8718,47 @@ output_ec_condition_for_handler (const enum cb_handler_type handler_type)
+		output_line("if (cob_glob_ptr->cob_exception_code == %d) "
+			"cob_fatal_exception (cob_glob_ptr->cob_exception_code);",
+			CB_EXCEPTION_CODE(COB_EC_SIZE_ZERO_DIVIDE));
+	}
```
can't you just have an integer `exception_code = 0` at the start of the function, then || the enabled exception values into it and then finally - if it isn't zero output one single exception check?

emilienlemaire replied:
We can't really have that, if the exception is not zero, it might still be an exception code that was disable. I've had the problem where I had `-fno-ec=size-overflow` yet it was still raised as other `ec-size` codes were enabled, and the check was a `||` (so if any was enabled, all were checked). This was the simplest way I could ensure that only the enabled exceptions are raised.

GitMensch replied:
let's go with it as you have it, along with the minimal -fno-ec in the NIST configuration - I can then have a look if my idea for making it "more direct" works after this code isupstream

GitMensch commented on cobc/codegen.c:
```diff
@@ -9151,6 +9192,15 @@ output_stmt (cb_tree x)
+
+		/* if ((!p->ex_handler || (p->ex_handler && p->handler_type != SIZE_ERROR_HANDLER))) {
+			output_size_exception_handler(p);
+		} */
```
if not relevant any more then please delete, otherwise move to `#if 0 /* comment why we may want that */`

GitMensch replied:
marked as resolved - then please commit your change solving that

GitMensch commented on cobc/parser.y:
```diff
@@ -18134,13 +18134,40 @@ on_size_error_phrases:
+		|| CB_EXCEPTION_ENABLE(COB_EC_SIZE_OVERFLOW)
+		|| CB_EXCEPTION_ENABLE(COB_EC_SIZE_TRUNCATION)
+		|| CB_EXCEPTION_ENABLE(COB_EC_SIZE_UNDERFLOW)
+		|| CB_EXCEPTION_ENABLE(COB_EC_SIZE_ZERO_DIVIDE))
```
style-only:

indent here and below: || by one space, not tab/multiple space, and ideally a space before the macro's (

GitMensch commented on cobc/parser.y:
```diff
@@ -18134,13 +18134,40 @@ on_size_error_phrases:
 	if ($2) {
 		cb_verify (cb_not_exception_before_exception,
 			_("NOT SIZE ERROR before SIZE ERROR"));
+	} else if ($1 && $1 != cb_int1) {
```
$1 here is always set as `not_on_size_error` is mandatory - you can drop the check for this and the setting to cb_int2 below

Nice and correct addition for that "null handler" in the parser BTW!


... but: it should not be added here by the standard (I think also by previous standards).
To quote: 

> If the EC-SIZE exception condition was raised and a SIZE ERROR phrase is not specified, processing
proceeds as specified in 14.6.13.1.3, Fatal exception conditions. If a NOT SIZE ERROR phrase is specified,
it is ignored.

The point here if one explicit codes `NOT SIZE ERROR` than this is identical to a coded `ON SIZE ERROR CONTINUE`

emilienlemaire replied:
I think this phrase means that if the exceptions exists and the `NOT SIZE ERROR` phrase is specified, then the `NOT SIZE ERROR` phrase is ignored, not that the exception is ignored

GitMensch replied:
I'm sure this isn't the case. You may try some computers, but the ones I've tested behave like my understanding of the standard. 
Logically: coding only NOT ON SIZE ERROR means you kind of expect but ignore one and execute the "important" part only if there isn't an error. 
Also note that coding it means (per standard) that the target field is unchanged if an error happens, otherwise it is undefined if no checking is on.

emilienlemaire replied:
Sure but the phrasing is `If a NOT SIZE ERROR phrase is specified, it is ignored.`, not `If a NOT SIZE ERROR phrase is specified, the exception is ignored.` so `it` must refer to the subject of the phrase, not something that was in a phrase before.

In addition this phrase appears after the description of what should happen in case of an exception without an `ON SIZE ERROR` phrase, and after specifically saying `and processing proceeds as specified in ...` so logically every steps the standard specify should happen before any consideration for `NOT ON SIZE ERROR`.

Afaik not many (if any) implementers have an exception system in their compiler, so maybe they work differently when having `NOT ON SIZE ERROR`, but if we decide to implement one, we should adhere to the standard as close as possible.

GitMensch replied:
let's take the following approach: we surround that with `#if 0 /* CHECKME: possible dialect configuration */` and get the code upstream, then check what implementations do and also with the standard committee what the intention is

GitMensch commented on libcob/call.c:
```diff
@@ -19,6 +19,7 @@
 */
 
 
+#include "common.h"
```
possibly added by clangd - can be dropped as it is explicit included via coblocal.h (and should always be included after config.h in any case)

GitMensch commented on tests/ChangeLog:
duplicated entries, can be dropped

GitMensch commented on tests/testsuite.src/data_display.at:
please make -fno-ec more specific here and in the other testcases as well

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -2689,8 +2689,10 @@ AT_DATA([prog.cob], [
-           COMPUTE Y = 99.  *> and once in a time this result was not stored
+           COMPUTE X = 100 ON SIZE ERROR CONTINUE.
+           *> this overflows which does not store a result
+           COMPUTE Y = 99 ON SIZE ERROR CONTINUE.
```
Y has no size error, so should not have that clause in

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -11733,7 +11737,7 @@ AT_DATA([prog.cob], [
 ])
 
-AT_CHECK([$COMPILE prog.cob], [0], [],
+AT_CHECK([$COMPILE -fno-ec=size-overflow prog.cob], [0], [],
```
With the ON SIZE above there is no need for -fno-ec here any more, is there?

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -15735,5 +15743,177 @@ AT_CHECK([$COBCRUN_DIRECT ./prog], [0],
+AT_SETUP([Enable EC-SIZE exceptions])
+AT_KEYWORDS([EXCEPTION EXCEPTION-CODE FATAL EC-SIZE])
+
+AT_DATA([prog.cob], [
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. prog.
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01 LARGE-VALUE PIC 9(3) VALUE 999.
+       01 SMALL-VALUE PIC 9(3) VALUE 1.
+       01 RESULT      PIC 9(3).
+       PROCEDURE DIVISION.
+           ADD SMALL-VALUE TO LARGE-VALUE
+                GIVING RESULT.
+           DISPLAY RESULT.
+           STOP RUN.
+])
+
+AT_CHECK([$COMPILE prog.cob], [0], [], [])
+AT_CHECK([$COBCRUN_DIRECT ./prog], [1],
+[], [libcob: prog.cob:10: error: Fatal exception not handled: EC-SIZE-OVERFLOW
+])
+
```
please add a prog0 here that is identical but has NOT ON SIZE ERROR clause - which should pass without an exception

GitMensch commented on libcob/ChangeLog:
```diff
@@ -15,6 +15,16 @@
+
+	* common.h, common.c: add `cob_exception_tab_fatal`, which informs
+	whether an exception is fatal or not
+	* common.h, common.c: add `cob_last_exception_fatal`, to check if the
+	last raised exception was fatal
+	* common.h, common.c: add enum member `COB_FERROR_FATAL_EC` for
+	`struct cob_fatal_error` which can be used when terminating the
+	program due to a fatal exception
```
style: indent by two spaces for continuation lines and  line-break before col81 (it looks like some words may wrap too early)

GitMensch commented on libcob/common.c:
```diff
@@ -468,6 +468,14 @@ static const int		cob_exception_tab_code[] = {
+#include "exception.def"
+	0		/* COB_EC_MAX */
+};
+
```
This and the related vars are not needed any more as there's now a new single function called, right?
So please revert the now not needed changes here and in common.h.

GitMensch commented on libcob/common.c:
```diff
@@ -2502,6 +2511,12 @@ cob_set_exception (const int id)
 }
 
+int
+cob_last_exception_fatal() {
```
`cob_last_exception_was_fatal (void)`, please - but that isn't used and needed any more, is it?

GitMensch reviewed (changes requested):
Just stumbled over an interesting explicit rule in COBOL2027 (before it possibly wasn't specified, would have to check): if the EC-SIZE happens on "item identification", then the ON SIZE phrase is not used but the fatal exception - if enabled should be done.

That is the part that will be covered by the "expression addition" - we should definitely add it to our testsuite - even with an expected failure:


```
MOVE 0 TO SOME
ADD TABLE-VAR (5 / SOME) TO ADD TABLE-VAR (5)
ON SIZE ERROR DISPLAY 'Should not be taken'.
```

The changes look good in general, but apart from some style and other minor issues there's one issue that is logically wrong - see the parser comment about NOT ON SIZE ERROR; I _guess_ that some of the NIST tests don't need the -fno-ec=size flags because of that - but this can be easily checked after adjusting parser.y

GitMensch commented on cobc/parser.y:
```diff
@@ -18176,6 +18203,7 @@ _not_on_size_error:
 not_on_size_error:
   NOT_SIZE_ERROR statement_list
   {
+	$$ = cb_int2;
```
as noted, this should not be needed, even if we enable the handler insertion as you've did

GitMensch reviewed (changes requested):
apart from the things mentioned in the PR this is good for upstream; please commit the missing adjustments and I can make it available

emilienlemaire committed:
Edits after review

emilienlemaire commented:
The PR should be OK to merge now, I should have edited everything you asked, and also fixed a few side effects those edits lit up, the main one being to reset the check of the exception before a new statement but still allow checking the exception with intrinsic functions (I checked on the standard, and the exception check should be reset between every statement, but the last exception code and stuff should still be available with intrinsics)

GitMensch commented on ABOUT-NLS:
this still needs to be reverted

emilienlemaire replied:
I thought I did :smiling_face_with_tear:, I'll add it with the fix on NC

GitMensch commented on tests/cobol85/report.pl:
```diff
@@ -236,6 +236,30 @@
+$cobc_flags{NC205A} = "-fno-ec=size-overflow";
+$cobc_flags{NC251A} = "-fno-ec=size-zero-divide";
+$cobc_flags{NC252A} = "-fno-ec=size-overflow";
+$cobc_flags{NC253A} = "-fno-ec=size-overflow";
```
Are all of those (after the change to NOT SIZE ERROR) still needed?
Are there some missing (as NIST NC currently fails in CI)?

emilienlemaire replied:
I tested again, some of them can be removed, but some errors occurs without a `NOT ON SIZE ERROR`, so we need to keep the one I updated in the last commit

GitMensch reviewed (changes requested):
Thanks. Currently NIST NC fails, but I guess you'll push a fix soon.

emilienlemaire commented:
Yes I'm working on a fix

emilienlemaire committed:
Fix NIST NC tests

GitMensch commented:
Something is broken. If we compile NC2032A with -fec=all (or just --debug) it aborts with

~~~
libcob: NC203A.CBL:307: error: Fatal exception not handled: EC-SIZE-ZERO-DIVIDE
libcob: NC203A.CBL:307: warning: implicit CLOSE of PRINT-FILE ('REPORT')

 Last statement of "NC203A" was ADD
        PASS OF SECT-NC203A-001 at NC203A.CBL:307
        DIV-TEST-F4-4-0 OF SECT-NC203A-001 at NC203A.CBL:483
        ENTRY NC203A at NC203A.CBL:295
 Started by ./NC203A
 ~~~
 
 The perform in 403 is

```cobol 
048300     DIVIDE  25COUNT INTO 100 GIVING 25ANS REMAINDER 25REM        NC2034.2
048400             ON SIZE ERROR                                        NC2034.2
048500             PERFORM PASS                                         NC2034.2   *>   here
048600             GO TO DIV-WRITE-F4-4-0.                              NC2034.2
```
 the ADD is 
```cobol
 030900 PASS.  MOVE "PASS " TO P-OR-F.  ADD 1 TO PASS-COUNTER.           NC2034.2
```

so the `DIVIDE` error which is correctly handled by its `ON SIZE ERROR` is evaluated witht the next statement `ADD`.

***Before any change in codegen it would good to have a similar example in the testsuite!***


The reason is that _before_ the statements that check the exceptions after their execution we need to reset it, which is done in codegen.c by those lines:

```c
		if (!p->file && (p->ex_handler || p->not_ex_handler)) {
			output_line ("cob_glob_ptr->cob_exception_code = 0;");
		}
```
which are not taken because no handler is set.

I suggest to add a check `p->handler_type == SIZE_ERROR_HANDLER` (+active ec-size) there and drop `cob_reset_exception`.

This should lead to much more (all?) NC tests not needing `-fno-ec=size`.


For the ec handler output I've meant something along

```c
static void
output_ec_size_handler (void)
{
	int ec_checked = 0;
	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_ADDRESS)) {
		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_ADDRESS);
	}
	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_EXPONENTIATION)) {
		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_EXPONENTIATION);
	}
	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_IMP)) {
		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_IMP);
	}
	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_OVERFLOW)) {
		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_OVERFLOW);
	}
	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_TRUNCATION)) {
		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_TRUNCATION);
	}
	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_UNDERFLOW)) {
		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_UNDERFLOW);
	}
	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_ZERO_DIVIDE)) {
		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_ZERO_DIVIDE);
	}
	if (ec_checked) {
		output_line("if ((cob_glob_ptr->cob_exception_code & 0x%04x) == 0x%04x)",
			CB_EXCEPTION_CODE (COB_EC_SIZE), CB_EXCEPTION_CODE (COB_EC_SIZE));
		output_line("\t" "cob_fatal_exception (cob_glob_ptr->cob_exception_code);");
	}
}
```

... but then possibly also have `cob_fatal_exception` not take an argument (it is only called if there is an error and operates on the last error).

emilienlemaire committed:
Fix exceptions wrongly raised in error handlers

emilienlemaire commented:
> For the ec handler output I've meant something along
> 
> ```c
> static void
> output_ec_size_handler (void)
> {
> 	int ec_checked = 0;
> 	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_ADDRESS)) {
> 		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_ADDRESS);
> 	}
> 	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_EXPONENTIATION)) {
> 		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_EXPONENTIATION);
> 	}
> 	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_IMP)) {
> 		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_IMP);
> 	}
> 	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_OVERFLOW)) {
> 		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_OVERFLOW);
> 	}
> 	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_TRUNCATION)) {
> 		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_TRUNCATION);
> 	}
> 	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_UNDERFLOW)) {
> 		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_UNDERFLOW);
> 	}
> 	if (CB_EXCEPTION_ENABLE (COB_EC_SIZE_ZERO_DIVIDE)) {
> 		ec_checked |= CB_EXCEPTION_CODE (COB_EC_SIZE_ZERO_DIVIDE);
> 	}
> 	if (ec_checked) {
> 		output_line("if ((cob_glob_ptr->cob_exception_code & 0x%04x) == 0x%04x)",
> 			CB_EXCEPTION_CODE (COB_EC_SIZE), CB_EXCEPTION_CODE (COB_EC_SIZE));
> 		output_line("\t" "cob_fatal_exception (cob_glob_ptr->cob_exception_code);");
> 	}
> }
> ```

On this, I was doing something close to this at one point, but let's say we have only `EC-SIZE-OVERFLOW` enabled, and we compute something that sets a `EC-SIZE-ZERO-DIVIDE`, which is set (among other places), with theses lines:
```c
	if (unlikely (mpz_sgn (d2->value) == 0)) {
		d1->scale = COB_DECIMAL_NAN;
		cob_set_exception (COB_EC_SIZE_ZERO_DIVIDE);
		return;
	}
```
then the generated condition will be true, and the exception will be wrongly raised, even if it was disabled.

As for the exceptions that were wrongly raised in handler statements, the last commit should fix it :smile:

emilienlemaire committed:
Fix some warnings

GitMensch commented:
> On this, I was doing something close to this at one point, but let's say we have only `EC-SIZE-OVERFLOW` enabled, and we compute something that sets a `EC-SIZE-ZERO-DIVIDE`, which is set (among other places), with theses lines:
> 
> ```c
> 	if (unlikely (mpz_sgn (d2->value) == 0)) {
> 		d1->scale = COB_DECIMAL_NAN;
> 		cob_set_exception (COB_EC_SIZE_ZERO_DIVIDE);
> 		return;
> 	}
> ```

Of course my code was wrong, should have been something like 

```c
	if (ec_checked) {
		output_line("if ((cob_glob_ptr->cob_exception_code & 0x%04x) == 0x%04x) && ((cob_glob_ptr->cob_exception_code & 0x%04x) != 0))",
			CB_EXCEPTION_CODE (COB_EC_SIZE), CB_EXCEPTION_CODE (COB_EC_SIZE)), (ec_checked & 0x00FF);
```

--> group is set, and one of the |= exception bits are set

emilienlemaire committed:
Better ec-size checking

emilienlemaire commented:
I implemented the check you suggested, but had to edit the exception codes in `exception.def` to avoid overlapping bits in the checks
